Update kotlinx-coroutines-test

dkhalanskyjb opened the pull request:
Fixes #1203
Fixes #1609 (I think)
Fixes #2379
Fixes #1749
Fixes #1204
Fixes #1390
Fixes #1222
Fixes #1395
Fixes #1881
Fixes #1910
Fixes #1772
Fixes #1626
Fixes #1742
Fixes #2082
Fixes #2102
Fixes #2405
Fixes #2462

qwwdfsad commented on kotlinx-coroutines-test/common/test/Helpers.kt:
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
```
It looks like you should update your licensing template in IDEA, the year should be 2021

dkhalanskyjb replied:
Dunno, I checked, and the configuration seems alright.

dkhalanskyjb commented on kotlinx-coroutines-test/common/src/TestCoroutineScope.kt:
```diff
@@ -62,11 +67,94 @@ private inline val CoroutineContext.uncaughtExceptionCaptor: UncaughtExceptionCa
+@Deprecated("The name of this function is misleading: it not only advances the time, but also runs the tasks " +
+    "scheduled *at* the ending moment.",
+    ReplaceWith("this.testScheduler.apply { advanceTimeBy(delayTimeMillis); runCurrent() }"),
+    DeprecationLevel.WARNING)
```
Fixed.

dkhalanskyjb commented on kotlinx-coroutines-test/common/src/TestCoroutineScope.kt:
```diff
@@ -10,47 +10,100 @@ import kotlin.coroutines.*
      */
-    public override fun cleanupTestCoroutines()
+    @ExperimentalCoroutinesApi
+    public fun cleanupTestCoroutines()
```
Fixed by deprecating `TestCoroutineScope`.

dkhalanskyjb commented on kotlinx-coroutines-test/common/src/TestCoroutineScope.kt:
```diff
@@ -10,47 +10,100 @@ import kotlin.coroutines.*
+    @ExperimentalCoroutinesApi
+    public fun cleanupTestCoroutines()
+
+    /**
```
Ditto.

dkhalanskyjb commented:
Since the last PR, I've rebased the changes and added https://github.com/Kotlin/kotlinx.coroutines/pull/2973/commits/1a8af52412824698df00b7e7336b7975bd04e3d3.

qwwdfsad commented on kotlinx-coroutines-test/common/src/TestCoroutineDispatchers.kt:
```diff
@@ -0,0 +1,158 @@
+ * @see UnconfinedTestDispatcher for a dispatcher that is not confined to any particular thread.
+ */
+@Suppress("FunctionName")
+public fun StandardTestDispatcher(
```
Should be experimental

qwwdfsad reviewed (approved):
Tremendeous and terrific work 🔥 

Please wait until https://github.com/Kotlin/kotlinx.coroutines/pull/3034 is merged, rebase on the recent `develop` and ensure that the build is green (I expect everything to work smoothly now), and then it's good to go!

radityagumay commented:
how do we know which version that will contains this fix?
